### Medir resultado do Active Learning

In [1]:
# Bibliotecas com métodos, formulas, rotinas para agilizar desenvolvimento
import pandas as pd
# Configurar o dataframe para exibir 131 colunas de um dataset
pd.set_option("max.columns", 131)

In [2]:
# Abrir o arquivo
df2 = pd.read_csv("active_labels.csv", index_col=0)

In [3]:
# Pegar somente a linhas cuja coluna Y é diferente de nula
df2 = df2[df2['y'].notnull()]

In [4]:
# Bibliotecas com métodos, formulas, rotinas para agilizar desenvolvimento
from sklearn.metrics import roc_auc_score, average_precision_score

In [5]:
# Exibir linha do dataset dentro do dataframe active_labels.csv ...
# Executar uma linha por vez
df2.query('y == 1 and p<=500').head() # ... Analisar o item 579 y == 1 e p = 375 ou 37,5%
df2.query('y == 0 and p>500').head() # ... Analisar o item 579 y == 1 e p = 826 ou 82,6%

watch-title  y  \
Column1                                                         
846      Kaggle Live-Coding: Graphing entity co-occurre...  0   
904              Kaggle Meetup: AdTracking Fraud Detection  0   
908      Kaggle Meetup: Quora Insincere Questions Class...  0   
919                            Kaggle Days Meetup Milan #1  0   
938                   Kaggle Meetup: Skin Cancer Diagnosis  0   

            watch-view-count                           watch-time-text  \
Column1                                                                  
846      448 visualizações  Transmitido ao vivo em 3 de ago. de 2018   
904      372 visualizações            Publicado em 1 de jun. de 2018   
908      399 visualizações            Publicado em 8 de mar. de 2019   
919      243 visualizações           Publicado em 22 de mai. de 2019   
938      663 visualizações            Publicado em 5 de out. de 2018   

        content_watch-info-tag-list  \
Column1                               
846           Ciência e tecnologia   
904           Ciência e tecnologia   
908           Ciência e tecnologia   
919                 Pessoas e blogs   
938           Ciência e tecnologia   

                                           watch7-headline  \
Column1                                                      
846      Kaggle Live-Coding: Graphing entity co-occurre...   
904              Kaggle Meetup: AdTracking Fraud Detection   
908      Kaggle Meetup: Quora Insincere Questions Class...   
919                            Kaggle Days Meetup Milan #1   
938                   Kaggle Meetup: Skin Cancer Diagnosis   

                  watch7-user-header watch8-sentiment-actions  \
Column1                                                         
846                           Kaggle            448 visualiza   
904      Learn Data ScienceCarregand            372 visualiza   
908      Learn Data ScienceCarregand            399 visualiza   
919           LogicAICarregando...              243 visualiza   
938      Learn Data ScienceCarregand            663 visualiza   

                                                  og:image  og:image:width  \
Column1                                                                      
846      https://i.ytimg.com/vi/iz-OrQLIlaI/maxresdefau...           12800   
904      https://i.ytimg.com/vi/kzkQTPeZPQ4/maxresdefau...           12800   
908      https://i.ytimg.com/vi/l7ZmviwSnXw/maxresdefau...           12800   
919       https://i.ytimg.com/vi/lkZET6thmLk/hqdefault.jpg            4800   
938      https://i.ytimg.com/vi/meg8I1GdtUg/maxresdefau...           12800   

         og:image:height                                     og:description  \
Column1                                                                       
846                 7200  Join Kaggle data scientist Rachael live as she...   
904                 7200  https://www.meetup.com/LearnDataScience/events...   
908                 7200  Kaggle page: https://www.kaggle.com/c/quora-in...   
919                 3600  15th of March had a place the first ever Kaggl...   
938                 7200  Video from this meetup: https://www.meetup.com...   

         og:video:width  og:video:height             og:video:tag  \
Column1                                                             
846             12800.0           7200.0                       CS   
904             12800.0           7200.0                      NaN   
908             12800.0           7200.0  artificial intelligence   
919             12800.0           7200.0                      NaN   
938             12800.0           7200.0         machine learning   

                            channel_link_0    p  
Column1                                          
846      /channel/UCSNeZleDn9c74yQc-EKnVTA  826  
904      /channel/UCJhW_16uxALr0X4olEW2p5A  518  
908      /channel/UCJhW_16uxALr0X4olEW2p5A  598  
919      /channel/UCZ_qlZbg9EzwRnLq_hFQumQ  691  
938      /channel/UCJh

In [6]:
# Executar as métricas average_precision_score, roc_auc_score
average_precision_score(df2['y'],df2['p']), roc_auc_score(df2['y'],df2['p'])

(0.26143174121317886, 0.6049606775559588)

#### Comparar as métricas
##### Métricas baseline
ap 0.14808785529715762 auc-roc 0.5705167173252279
##### Métricas atuais, sem entrar em pormenores
ap 0.26143174121317886 auc-roc 0.6049606775559588

In [7]:
# Criar uma nova coluna Novo no dataframe 
df2['novo'] = 1

In [8]:
# Exibir quantidade de colunas e linhas, nome, tipo e quantidade do conteudo de cada coluna, memória usada
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 106 entries, 505 to 1052
Data columns (total 18 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   watch-title                  106 non-null    object 
 1   y                            106 non-null    int64  
 2   watch-view-count             106 non-null    object 
 3   watch-time-text              106 non-null    object 
 4   content_watch-info-tag-list  106 non-null    object 
 5   watch7-headline              106 non-null    object 
 6   watch7-user-header           106 non-null    object 
 7   watch8-sentiment-actions     106 non-null    object 
 8   og:image                     106 non-null    object 
 9   og:image:width               106 non-null    int64  
 10  og:image:height              106 non-null    int64  
 11  og:description               106 non-null    object 
 12  og:video:width               105 non-null    float64
 13  og:video:height  

In [9]:
# Abrir o arquivo
df1 = pd.read_csv("raw_data_with_labels.csv", index_col=0)

In [10]:
# Pegar somente a linhas cuja coluna Y é diferente de nula
df1 = df1[df1['y'].notnull()]

In [11]:
# Exibir as primeiras cinco linhas do dataset dentro do dataframe
df1.head()

watch-title    y  \
0  How to Become A Machine Learning Engineer | Ho...  0.0   
1  BLOOPERS - Behind The Scenes. | DATA SCIENCE x...  0.0   
2  Michael I. Jordan: Machine Learning: Dynamical...  1.0   
3              Best Deep Learning Tools - Welcome.AI  0.0   
4  Kaggle Live-Coding: RNNs for Sarcasm Detection...  1.0   

       watch-view-count                            watch-time-text  \
0  28.028 visualizações             Publicado em 3 de set. de 2018   
1   1.131 visualizações            Publicado em 16 de nov. de 2018   
2   1.816 visualizações             Publicado em 2 de mai. de 2019   
3   1.171 visualizações            Publicado em 13 de ago. de 2019   
4   1.228 visualizações  Transmitido ao vivo em 30 de nov. de 2018   

                         content_watch-info-tag-list  \
0                                           Educação   
1                                    Pessoas e blogs   
2  Licença de atribuição Creative Commons (reutil...   
3                               Ciência e tecnologia   
4                               Ciência e tecnologia   

                                     watch7-headline  \
0  #MachineLearningAlgorithms #Datasciencecourse ...   
1  #FAIL #insidezalando\n\n\n\n    BLOOPERS - Beh...   
2  #purdue #michaelijordan #engineering\n\n\n\n  ...   
3              Best Deep Learning Tools - Welcome.AI   
4  Kaggle Live-Coding: RNNs for Sarcasm Detection...   

                                  watch7-user-header  \
0  Simplilearn\n\n\n\n\n\n\n\n\n\n\n\n\n\n\nCarre...   
1  Inside Zalando\n\n\n\n\n\n\n\n\n\n\n\n\n\nCarr...   
2  Purdue Engineering\n\n\n\n\n\n\n\n\n\n\n\n\n\n...   
3  Welcome.AI\n\n\n\n\n\n\n\n\n\n\n\n\n\nCarregan...   
4  Kaggle\n\n\n\n\n\n\n\n\n\n\n\n\n\nCarregando.....   

                            watch8-sentiment-actions  \
0  28.028 visualizações\n\n\n\n\n\n\n\n601\n\nGos...   
1  1.131 visualizações\n\n\n\n\n\n\n\n20\n\nGosto...   
2  1.816 visualizações\n\n\n\n\n\n\n\n42\n\nGosto...   
3  1.171 visualizações\n\n\n\n\n\n\n\n14\n\nGosto...   
4  1.228 visualizações\n\n\n\n\n\n\n\n28\n\nGosto...   

                                            og:image  og:image:width  \
0  https://i.ytimg.com/vi/-5hEYRt8JE0/maxresdefau...          1280.0   
1  https://i.ytimg.com/vi/-7GiiT0yEyk/maxresdefau...          1280.0   
2  https://i.ytimg.com/vi/-8yYFdV5SOc/maxresdefau...          1280.0   
3  https://i.ytimg.com/vi/-9LLrwW1Vdo/maxresdefau...          1280.0   
4  https://i.ytimg.com/vi/-9U84J178OQ/maxresdefau...          1280.0   

   og:image:height                                     og:description  \
0            720.0  This video on "How to become a Machine Learnin...   
1            720.0  #FAIL :) Have fun - and join our teams: https:...   
2            720.0  2019 Purdue Engineering Distinguished Lecture ...   
3            720.0  A collection of the 5 best deep learning tools...   
4            720.0  Join Kaggle data scientist Rachael live as she...   

   og:video:width  og:video:height         og:video:tag  \
0          1280.0            720.0          simplilearn   
1          1280.0            720.0    employer branding   
2          1280.0            720.0  electrical engineer   
3          1280.0            720.0               Watson   
4          1280.0            720.0                   CS   

                      channel_link_0  
0  /channel/UCsvqVGtbbyHaMoevxPAq9Fg  
1  /channel/UCTPin8TK-KRSI9zo9FoxG0g  
2  /channel/UC8FZ6dzFVkCACLH9YoMNFog  
3  /channel/UC_215Y7rOAsqnFkO_hnpdIg  
4  /channel/UCSNeZleDn9c74yQc-EKnVTA

In [12]:
# Exibir quantidade de colunas e linhas, nome, tipo e quantidade do conteudo de cada coluna, memória usada
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 498 entries, 0 to 500
Data columns (total 16 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   watch-title                  498 non-null    object 
 1   y                            498 non-null    float64
 2   watch-view-count             498 non-null    object 
 3   watch-time-text              498 non-null    object 
 4   content_watch-info-tag-list  498 non-null    object 
 5   watch7-headline              498 non-null    object 
 6   watch7-user-header           498 non-null    object 
 7   watch8-sentiment-actions     498 non-null    object 
 8   og:image                     498 non-null    object 
 9   og:image:width               498 non-null    float64
 10  og:image:height              498 non-null    float64
 11  og:description               498 non-null    object 
 12  og:video:width               495 non-null    float64
 13  og:video:height     

In [13]:
# Concatenar os dataframes dificeis e aleatorios
# Atentar que o dataframe df1 recebera a coluna Novo com conteudo nan. Isto significa que dever ser limpo
df = pd.concat([df1, df2.drop("p", axis=1)])

#### Limpar e/ou tratar os dados para melhor uso no modelo de predição

In [14]:
# Criar um dataframe apenas com indices iguais ao dataframe de concatenação
# Este servirá para popularmos com as colunas limpas e tratadas
df_limpo = pd.DataFrame(index=df.index)

In [15]:
# Criar colunas title e novo neste novo dataframe df_limpo
# Extrair a descrição dos títulos
df_limpo['title'] = df['watch-title']
# Eliminar o conteúdo do campo nan ao preenche-lo com 0
df_limpo['novo'] = df['novo'].fillna(0)

#### Extrair apenas a data (dia, mês e ano) de uma coluna com strings

In [16]:
# Extrair os numeros que representam data de publicação ' dia (numero), mes (string) e ano (numero)'
clean_date = df['watch-time-text'].str.extract(r"(\d+) de ([a-z]+)\. de (\d+)")
# Incluir zero à frente do numero abaixo de 10 que representa do dia 
clean_date[0] = clean_date[0].map(lambda x: "0"+x[0] if len(x) == 1 else x)
#clean_date[1] = clean_date[1].map(lambda x: x[0].upper()+x[1:])

In [17]:
# Transformar mes que esta portugues para mes em inglês
mapa_meses = {"jan": "Jan",
              "fev": "Feb",
              "mar": "Mar", 
              "abr": "Apr", 
              "mai": "May", 
              "jun": "Jun",
              "jul": "Jul",
              "ago": "Aug", 
              "set": "Sep", 
              "out": "Oct", 
              "nov": "Nov",
              "dez": "Dec"}
# Transforma os meses
clean_date[1] = clean_date[1].map(mapa_meses)

In [18]:
# Unir numeros que representam data ' dia (numero), mes (string) e ano (numero)'
clean_date = clean_date.apply(lambda x: " ".join(x), axis=1)
# Converter para data os dados extraidos em clean_date e gravar no campo date criado no dataframe df_limpo
df_limpo['date'] = pd.to_datetime(clean_date, format="%d %b %Y")

In [19]:
# Exibir cinco primeiras linhas do dataset dentro do dataframe
clean_date.head()

0    03 Sep 2018
1    16 Nov 2018
2    02 May 2019
3    13 Aug 2019
4    30 Nov 2018
dtype: object

#### Extrair apenas a quantidade de uma coluna com strings
#### Eliminar o conteúdo nan

In [20]:
# Extrair apenas o número, eliminar o ponto que esta no numero, eliminar o tipo nan da coluna e converter
# todo o conteudo para o tipo integer
# expand=False faz o pandas retornar apenas série ao invés de dataframe
views = df['watch-view-count'].str.extract(r"(\d+\.?\d*)", expand=False).str.replace(".", "").fillna(0).astype(int)
# Criar coluna views no dataframe df_limpo e popular com os novos dados numéricos
df_limpo['views'] = views

#### Criar Features

In [21]:
# Criar dataframe features para trabalhar melhor nas features sob o dataframe de concatenação df_limpo
features = pd.DataFrame(index=df_limpo.index)
# Copiar a coluna y do dataframe de concatenação dos dataframes df1 e df2
y = df['y'].copy()

In [22]:
# Bibliotecas com métodos, formulas, rotinas para agilizar desenvolvimento
import numpy as np

In [23]:
# Subtrair da data da publicação do vídeo uma data fixa posterior, como parâmetro fixo, depois
# converter para uma estrutura série de float timedelta64 do numpy
features['tempo_desde_pub'] = (pd.to_datetime("2019-12-03") - df_limpo['date']) / np.timedelta64(1, 'D')
# Apenas copiar os dados da coluna views, não precisam de tratamento
features['views'] = df_limpo['views']
# Dividir o tempo de publicação do vídeo pela quantidade de views e gravar na nova coluna views_por_dia
# para ter uma quantidade acumulada aproximada de views por dia
features['views_por_dia'] = features['views'] / features['tempo_desde_pub']
# Apagar a coluna tempo_desde_pub, para exibir melhor as informações em gráfico
features = features.drop(['tempo_desde_pub'], axis=1)

In [24]:
features.head()

views  views_por_dia
0  28028      61.464912
1   1131       2.960733
2   1816       8.446512
3   1171      10.455357
4   1228       3.336957

#### AUMENTAR QUANTIDADE DADOS DE VALIDAÇÃO

In [25]:
# No dataset treino os dados foi aplicado a a validação temporal por ser uma time series, sendo 
# separados aproximadamente 50% treino e 50% teste (ao invés de utilizar o recurso de percentual 
# de métodos automáticos de separação)
# A variável mask_train conterá apenas os aproximadamente 500 registros antigos
# dos 100 exemplos que haviam sido concatenados em pd.concat([...])
mask_train = (df_limpo['date'] < "2019-04-01") & (df_limpo['novo'] == 0)
mask_val = (df_limpo['date'] >= "2019-04-01")
Xtrain, Xval = features[mask_train], features[mask_val]
ytrain, yval = y[mask_train], y[mask_val]

In [26]:
# Exibir a quantidade de vídeos separados para treino e teste
Xtrain.shape, Xval.shape, ytrain.shape, yval.shape

((228, 2), (319, 2), (228,), (319,))

In [27]:
# Bibliotecas com métodos, formulas, rotinas para agilizar desenvolvimento
from sklearn.feature_extraction.text import TfidfVectorizer

In [28]:
# Separar as descrições dos titulos para treino e teste
title_train = df_limpo[mask_train]['title']
title_val = df_limpo[mask_val]['title']

In [29]:
# Instanciar o objeto com seu min_df=2. Este é o número mínimo que uma palavra precisa aparecer para se 
# tornar uma coluna. Se min_df=1 gera 742 colunas, igual a 2 gera 193 colunas por que a maior parte das 
# palavras so aparece em um vídeo
title_vec = TfidfVectorizer(min_df=2)
#Transformar textos dos vídeos que estão na coluna watch-title em uma representação significante de números
# fit_transform armazena as palavras que detectou e armazena a frequencia de vezes que as detectou
# no dataset, para multiplicar na fórmula do Tfidf
# Bag of words em uma sparse matrix
title_bow_train = title_vec.fit_transform(title_train)
# Transformar com base nas palavras que aprendeu utilizando o dataset de treino
title_bow_val = title_vec.transform(title_val)

In [30]:
# Exibir a forma do dataset treinado : 228 linhas, 193 colunas
title_bow_train.shape

(228, 193)

In [31]:
# Bibliotecas com métodos, formulas, rotinas para agilizar desenvolvimento
from scipy.sparse import hstack, vstack

In [32]:
# Concatenar variáveis com dados numéricos com variáveis com dados de texto
# 228 * 193 matriz sparsa e x 228 * 2 features numericas a dimensao depois de juntas vao para 195 colunas
Xtrain_wtitle = hstack([Xtrain, title_bow_train])
Xval_wtitle = hstack([Xval, title_bow_val])

In [33]:
# Exibir quantidade de linhas e colunas
Xtrain_wtitle.shape, Xval_wtitle.shape

((228, 195), (319, 195))

In [34]:
# Bibliotecas com métodos, formulas, rotinas para agilizar desenvolvimento
from sklearn.ensemble import RandomForestClassifier

In [35]:
# Instânciar o objeto RandomForestClassifier : 1000 arvores,class_weight=balancear o entre as variáveis 
# train e test
mdl = RandomForestClassifier(n_estimators=1000, random_state=0, class_weight="balanced", n_jobs=6)
mdl.fit(Xtrain_wtitle, ytrain)

RandomForestClassifier(class_weight='balanced', n_estimators=1000, n_jobs=6,
                       random_state=0)

In [36]:
# Probabilidade de acerto da predição
p = mdl.predict_proba(Xval_wtitle)[:, 1]

In [37]:
from sklearn.metrics import roc_auc_score, average_precision_score

In [38]:
average_precision_score(yval, p), roc_auc_score(yval, p)

(0.2346888243978529, 0.6120799490364708)

ANTES
ap 0.1918043901336543, auc 0.5848024316109421 - mindf=2
##### A precisão da pontuação entre a probabilidade dos dados de treino e dados de validação infelizmente diminuiu mas o auc melhorou -esta mais próximo de 1. Entendemos que ambos estão bem sensíveis, por conta da pouca quantidade de dados.

#### AUMENTAR QUANTIDADE DADOS DE TREINO

In [39]:
# No dataset treino foi aplicado a validação temporal por ser uma time series, e esta sendo considerado
# os +- 500 exemplos originais + os 100 novos exemplos concatenados
# Na mask_train foram retirados pela instrução (df_limpo['novo'] == 0) todos os registros (linhas+campos) 
# dos 100 exemplos que haviam sido concatenados em pd.concat([...])
mask_train = (df_limpo['date'] < "2019-04-01")
mask_val = (df_limpo['date'] >= "2019-04-01") & (df_limpo['novo'] == 0)
Xtrain, Xval = features[mask_train], features[mask_val]
ytrain, yval = y[mask_train], y[mask_val]
#mask_val = (df_limpo['date'] >= "2019-04-01")

In [40]:
# Exibir quantidade de linhas e colunas
Xtrain.shape, Xval.shape, ytrain.shape, yval.shape

((285, 2), (270, 2), (285,), (270,))

In [41]:
# Separar as descrições dos titulos para treino e teste
title_train = df_limpo[mask_train]['title']
title_val = df_limpo[mask_val]['title']

In [42]:
# Instanciar o objeto com seu min_df=2. Este é o número mínimo que uma palavra precisa aparecer para se 
# tornar uma coluna. Se min_df=1 gera 742 colunas, igual a 2 gera 193 colunas por que a maior parte das 
# palavras so aparece em um vídeo
title_vec = TfidfVectorizer(min_df=2)

In [43]:
#Transformar textos dos vídeos que estão na coluna watch-title em uma representação significante de números
# fit_transform armazena as palavras que detectou e armazena a frequencia de vezes que as detectou
# no dataset, para multiplicar na fórmula do Tfidf
# Bag of words em uma sparse matrix
title_bow_train = title_vec.fit_transform(title_train)
title_bow_val = title_vec.transform(title_val)

In [44]:
# Exibir a forma do dataset treinado : 228 linhas, 193 colunas
title_bow_train.shape, title_bow_val.shape

((285, 236), (270, 236))

In [45]:
# Concatenar variáveis com dados numéricos com variáveis com dados de texto
# 282 * 2 matriz sparsa e x 282 * 233 features numericas a dimensao depois de juntas vao para 235 colunas
Xtrain_wtitle = hstack([Xtrain, title_bow_train])
# 270 * 2 matriz sparsa e x 270 * 233 features numericas a dimensao depois de juntas vao para 235 colunas
Xval_wtitle = hstack([Xval, title_bow_val])

In [46]:
# Instânciar o objeto RandomForestClassifier : 1000 arvores,class_weight=balancear o entre as variáveis 
# train e test
mdl = RandomForestClassifier(n_estimators=1000, random_state=0, class_weight="balanced", n_jobs=6)
mdl.fit(Xtrain_wtitle, ytrain)

RandomForestClassifier(class_weight='balanced', n_estimators=1000, n_jobs=6,
                       random_state=0)

In [47]:
# Probabilidade de acerto da predição
p = mdl.predict_proba(Xval_wtitle)[:, 1]

In [48]:
# Analisar a precisão da predição do dataset de test. Este melhor que o modelo 1 cuja nossa baseline com
# o modelo DecisionTreeClassifier era 0.148087
average_precision_score(yval, p), roc_auc_score(yval, p)

(0.17126455511050453, 0.5700911854103343)

#### UTILIZAR DADOS DE TREINO E VALIDAÇÃO NOVOS

In [49]:
# No dataset treino foi aplicado a validação temporal por ser uma time series, e esta sendo considerado
# os +- 500 exemplos originais + os 100 novos exemplos concatenados
# Na mask_train foram retirados pela instrução (df_limpo['novo'] == 0) todos os registros (linhas+campos) 
# dos 100 exemplos que haviam sido concatenados em pd.concat([...])
mask_train = (df_limpo['date'] < "2019-04-01")
mask_val = (df_limpo['date'] >= "2019-04-01")
Xtrain, Xval = features[mask_train], features[mask_val]
ytrain, yval = y[mask_train], y[mask_val]

In [50]:
# Exibir quantidade de linhas e colunas
Xtrain.shape, Xval.shape, ytrain.shape, yval.shape

((285, 2), (319, 2), (285,), (319,))

In [51]:
# Separar as descrições dos titulos para treino e teste
title_train = df_limpo[mask_train]['title']
title_val = df_limpo[mask_val]['title']

In [52]:
# Instanciar o objeto com seu min_df=2. Este é o número mínimo que uma palavra precisa aparecer para se 
# tornar uma coluna. Se min_df=1 gera 742 colunas, igual a 2 gera 193 colunas por que a maior parte das 
# palavras so aparece em um vídeo
title_vec = TfidfVectorizer(min_df=2)

In [53]:
#Transformar textos dos vídeos que estão na coluna watch-title em uma representação significante de números
# fit_transform armazena as palavras que detectou e armazena a frequencia de vezes que as detectou
# no dataset, para multiplicar na fórmula do Tfidf
# Bag of words em uma sparse matrix
title_bow_train = title_vec.fit_transform(title_train)
title_bow_val = title_vec.transform(title_val)

In [54]:
# Exibir a forma do dataset treinado : 282 linhas, 233 colunas
title_bow_train.shape, title_bow_val.shape

((285, 236), (319, 236))

In [55]:
# Concatenar variáveis com dados numéricos com variáveis com dados de texto
# 282 * 2 matriz sparsa e x 282 * 233 features numericas a dimensao depois de juntas vao para 235 colunas
Xtrain_wtitle = hstack([Xtrain, title_bow_train])
# 270 * 2 matriz sparsa e x 270 * 233 features numericas a dimensao depois de juntas vao para 235 colunas
Xval_wtitle = hstack([Xval, title_bow_val])

In [56]:
# Instânciar o objeto RandomForestClassifier : 1000 arvores,class_weight=balancear o entre as variáveis 
# train e test
mdl = RandomForestClassifier(n_estimators=1000, random_state=0, class_weight="balanced", n_jobs=6)
mdl.fit(Xtrain_wtitle, ytrain)

RandomForestClassifier(class_weight='balanced', n_estimators=1000, n_jobs=6,
                       random_state=0)

In [57]:
# Probabilidade de acerto da predição
p = mdl.predict_proba(Xval_wtitle)[:, 1]

In [58]:
# Analisar a precisão da predição do dataset de test. Este melhor que o modelo 1 cuja nossa baseline com
# o modelo DecisionTreeClassifier era 0.148087
average_precision_score(yval, p), roc_auc_score(yval, p)

(0.19172255241322184, 0.5873944895684026)

##### R E S U L T A D O S
ANTES - Gerado 2-Random-Forest-Classifier.md<br>
ap 0.1918043901336543     auc 0.5848024316109421 - mindf=2<br>
ap 0.19493368357443233    auc 0.59580547112462 - mindf=2<br>

Validação nova e Treino antigo<br>
ap 0.1872802752830275, auc 0.5935436218282933<br>

Validação antiga e Treino novo<br>
ap 0.17697903398635972, auc 0.5964133738601823<br>

Validação nova e Treino novo<br>
ap 0.19342838733336654, auc 0.6095324991310394<br>
ap 0.18913174974061997, auc 0.6072297532151547 (date 20200830)